File: 05.raspberry pi/01.OpenAPI/ex03.py
from requests import get
import json

API_KEY = 'c2bf55882617af0dce37b001fcb64dcc'

def get_weather(city='Seoul'):
    URL = f'http://api.openweathermap.org/data/2.5/weather?q={city}&appid={API_KEY}&lang=kr'
    print(URL)
    weather = {}
    res = get(URL)
    if res.status_code == 200:
        result = res.json()
        weather['main'] = result['weather'][0]['main']
        weather['description'] = result['weather'][0]['description']
        # print(result['weather'][0]['description'])
        icon = result['weather'][0]['icon']
        weather['icon'] = f'http://openweathermap.org/img/w/{icon}.png'
        weather['etc'] = result['main']
    else:
        print('error', res.status_code)

    return weather

def download(url, file_name):
    with open(file_name, "wb") as file:
        response = get(url)
        file.write(response.content)

weather = get_weather()
print(json.dumps(weather, indent=4, ensure_ascii=False)) # utf8 사용
print("섭씨 온도 :", weather['etc']['temp'] - 273.15)
download(weather['icon'], weather['icon'].split('/')[-1])

File: 03.web/05.javascript/09. Ajax/ex03.js
// 날씨 받아오기 - async, await

function showWeather(w) {
  const main = w.main;
  const url_img = `http://api.openweathermap.org/img/w/${w.weather[0].icon}.png`;
  const template = `
    <h3>${w.name}</h3>
    <p>
      <img src="${url_img}" />
      ${w.weather[0].description}
    </p>
    <p>기온 : ${main.temp}</p>
    <p>습도 : ${main.humidity}</p>
    <p>풍속 : ${w.wind.speed} m/s, 풍향 : ${w.wind.deg}°</p>
  `;
  
  const weather = document.getElementById('weather');
  weather.innerHTML = template;
}

const getWeather = async (city) => {
  const url = 'http://api.openweathermap.org/data/2.5/weather';
  const lang = 'kr';
  const units = "metric";
  const params = { q: city, appid: weather_key, lang, units };

  try {
    const res = await axios.get(url, { params });
    showWeather(res.data);
  } catch(e) {
    console.log(e);
  }
};

const getCityWeather = e => {
  console.log(e.target);
  const city = e.target.dataset.name;
  getWeather(city);
};

document.addEventListener('DOMContentLoaded', function() {
  // 이 코드는 비효율적. DOM이 많아지면 브라우저가 느려질 수 있음
  // let cities = document.getElementsByClassName('city');
  // for (let i = 0; i < cities.length; ++i) {
  //   cities[i].onclick = getWeather;
  // }
  document.getElementById('cities').onclick = getCityWeather; // 이벤트 위임(delegation)
}, false);
